std/shift.asm

```diff
@@ -14,14 +15,19 @@ machine Shift(latch, operation_id) {
     col fixed FACTOR_ROW(i) { (i + 1) % 4 };
     col fixed FACTOR(i) { 1 << (((i + 1) % 4) * 8) };
 
-    col fixed P_A(i) { i % 256 };
-    col fixed P_B(i) { (i / 256) % 32 };
-    col fixed P_ROW(i) { (i / (256 * 32)) % 4 };
-    col fixed P_operation(i) { (i / (256 * 32 * 4)) % 2 };
+    let inputs = cross_product([256, 32, 4, 2]);
+    let a: int -> int = inputs[0];
+    let b: int -> int = inputs[1];
+    let row: int -> int = inputs[2];
+    let op: int -> int = inputs[3];
+    let P_A: col = a;
```

Comment:
how does this conversion work, if `P_A` needs to return a `fe`?

Reply:
That's one of the very few leftover implicit conversion rules: `col` is not really a type, it is more like a marker. We could also say that we go back to `col P_A;` to declare witness columns and `col P_A = ...` for fixed columns.

Having said that:

If you declare a symbol of type `col`, then the value of that declaration can either be `int -> int` or `int -> fe`.

If you reference a symbol of type `col`, the type of that reference will be `expr`.

Reply:
I think `let ... col` is better